File: tests/sdk/go/dynamodb_lsi_partiql_test.go
package sdk_test

import (
	"context"
	"testing"

	"github.com/aws/aws-sdk-go-v2/aws"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb"
	"github.com/aws/aws-sdk-go-v2/service/dynamodb/types"
)

func TestDynamoDBLSIQuery(t *testing.T) {
	requireReady(t)
	cfg := loadAWSConfig(t)
	client := newDDB(t, cfg)
	ctx := context.Background()

	table := uniquePrefix(t) + "-lsi"
	_, err := client.CreateTable(ctx, &dynamodb.CreateTableInput{
		TableName: aws.String(table),
		AttributeDefinitions: []types.AttributeDefinition{
			{AttributeName: aws.String("pk"), AttributeType: types.ScalarAttributeTypeS},
			{AttributeName: aws.String("sk"), AttributeType: types.ScalarAttributeTypeS},
			{AttributeName: aws.String("status"), AttributeType: types.ScalarAttributeTypeS},
		},
		KeySchema: []types.KeySchemaElement{
			{AttributeName: aws.String("pk"), KeyType: types.KeyTypeHash},
			{AttributeName: aws.String("sk"), KeyType: types.KeyTypeRange},
		},
		LocalSecondaryIndexes: []types.LocalSecondaryIndex{
			{
				IndexName: aws.String("ByStatus"),
				KeySchema: []types.KeySchemaElement{
					{AttributeName: aws.String("pk"), KeyType: types.KeyTypeHash},
					{AttributeName: aws.String("status"), KeyType: types.KeyTypeRange},
				},
				Projection: &types.Projection{ProjectionType: types.ProjectionTypeAll},
			},
		},
		BillingMode: types.BillingModePayPerRequest,
	})
	if err != nil {
		t.Fatalf("CreateTable: %v", err)
	}
	t.Cleanup(func() {
		_, _ = client.DeleteTable(ctx, &dynamodb.DeleteTableInput{TableName: aws.String(table)})
	})

	desc, err := client.DescribeTable(ctx, &dynamodb.DescribeTableInput{TableName: aws.String(table)})
	if err != nil {
		t.Fatalf("DescribeTable: %v", err)
	}
	if desc.Table == nil || len(desc.Table.LocalSecondaryIndexes) != 1 {
		t.Fatalf("LSI missing: %+v", desc.Table)
	}

	_, err = client.PutItem(ctx, &dynamodb.PutItemInput{
		TableName: aws.String(table),
		Item: map[string]types.AttributeValue{
			"pk":     &types.AttributeValueMemberS{Value: "u1"},
			"sk":     &types.AttributeValueMemberS{Value: "o1"},
			"status": &types.AttributeValueMemberS{Value: "OPEN"},
		},
	})
	if err != nil {
		t.Fatalf("PutItem: %v", err)
	}

	out, err := client.Query(ctx, &dynamodb.QueryInput{
		TableName:              aws.String(table),
		IndexName:              aws.String("ByStatus"),
		KeyConditionExpression: aws.String("pk = :pk AND #s = :st"),
		ExpressionAttributeNames: map[string]string{
			"#s": "status",
		},
		ExpressionAttributeValues: map[string]types.AttributeValue{
			":pk": &types.AttributeValueMemberS{Value: "u1"},
			":st": &types.AttributeValueMemberS{Value: "OPEN"},
		},
	})
	if err != nil {
		t.Fatalf("Query LSI: %v", err)
	}
	if len(out.Items) != 1 {
		t.Fatalf("items=%d", len(out.Items))
	}
	pk, ok := out.Items[0]["pk"].(*types.AttributeValueMemberS)
	if !ok || pk.Value != "u1" {
		t.Fatalf("pk=%v", out.Items[0]["pk"])
	}
	st, ok := out.Items[0]["status"].(*types.AttributeValueMemberS)
	if !ok || st.Value != "OPEN" {
		t.Fatalf("status=%v", out.Items[0]["status"])
	}
}

func TestDynamoDBPartiQLStatement(t *testing.T) {
	requireReady(t)
	cfg := loadAWSConfig(t)
	client := newDDB(t, cfg)
	ctx := context.Background()

	table := uniquePrefix(t) + "-partiql"
	_, err := client.CreateTable(ctx, &dynamodb.CreateTableInput{
		TableName: aws.String(table),
		AttributeDefinitions: []types.AttributeDefinition{
			{AttributeName: aws.String("pk"), AttributeType: types.ScalarAttributeTypeS},
		},
		KeySchema: []types.KeySchemaElement{
			{AttributeName: aws.String("pk"), KeyType: types.KeyTypeHash},
		},
		BillingMode: types.BillingModePayPerRequest,
	})
	if err != nil {
		t.Fatalf("CreateTable: %v", err)
	}
	t.Cleanup(func() {
		_, _ = client.DeleteTable(ctx, &dynamodb.DeleteTableInput{TableName: aws.String(table)})
	})

	_, err = client.ExecuteStatement(ctx, &dynamodb.ExecuteStatementInput{
		Statement: aws.String("INSERT INTO \"" + table + "\" VALUE {'pk':?,'data':?}"),
		Parameters: []types.AttributeValue{
			&types.AttributeValueMemberS{Value: "1"},
			&types.AttributeValueMemberS{Value: "partiql"},
		},
	})
	if err != nil {
		t.Fatalf("ExecuteStatement INSERT: %v", err)
	}

	sel, err := client.ExecuteStatement(ctx, &dynamodb.ExecuteStatementInput{
		Statement: aws.String("SELECT * FROM \"" + table + "\" WHERE pk=?"),
		Parameters: []types.AttributeValue{
			&types.AttributeValueMemberS{Value: "1"},
		},
	})
	if err != nil {
		t.Fatalf("ExecuteStatement SELECT: %v", err)
	}
	if len(sel.Items) != 1 {
		t.Fatalf("select items=%d", len(sel.Items))
	}
	data, ok := sel.Items[0]["data"].(*types.AttributeValueMemberS)
	if !ok || data.Value != "partiql" {
		t.Fatalf("data=%v", sel.Items[0]["data"])
	}
}

File: tests/sdk/nodejs/test/dynamodb-lsi-partiql.test.mjs
import { test } from "node:test";
import assert from "node:assert/strict";
import {
  CreateTableCommand,
  DeleteTableCommand,
  DescribeTableCommand,
  ExecuteStatementCommand,
  PutItemCommand,
  QueryCommand,
} from "@aws-sdk/client-dynamodb";
import { newDDB, requireReady, uniquePrefix } from "../lib/helpers.mjs";

test("DynamoDB LSI query", async (t) => {
  if (!(await requireReady(t))) return;
  const client = newDDB();
  const table = `${uniquePrefix()}-lsi`;

  await client.send(
    new CreateTableCommand({
      TableName: table,
      AttributeDefinitions: [
        { AttributeName: "pk", AttributeType: "S" },
        { AttributeName: "sk", AttributeType: "S" },
        { AttributeName: "status", AttributeType: "S" },
      ],
      KeySchema: [
        { AttributeName: "pk", KeyType: "HASH" },
        { AttributeName: "sk", KeyType: "RANGE" },
      ],
      LocalSecondaryIndexes: [
        {
          IndexName: "ByStatus",
          KeySchema: [
            { AttributeName: "pk", KeyType: "HASH" },
            { AttributeName: "status", KeyType: "RANGE" },
          ],
          Projection: { ProjectionType: "ALL" },
        },
      ],
      BillingMode: "PAY_PER_REQUEST",
    }),
  );
  t.after(async () => {
    try {
      await client.send(new DeleteTableCommand({ TableName: table }));
    } catch {
      /* ignore */
    }
  });

  const desc = await client.send(new DescribeTableCommand({ TableName: table }));
  assert.equal(desc.Table?.LocalSecondaryIndexes?.length, 1);

  await client.send(
    new PutItemCommand({
      TableName: table,
      Item: {
        pk: { S: "u1" },
        sk: { S: "o1" },
        status: { S: "OPEN" },
      },
    }),
  );

  const out = await client.send(
    new QueryCommand({
      TableName: table,
      IndexName: "ByStatus",
      KeyConditionExpression: "pk = :pk AND #s = :st",
      ExpressionAttributeNames: { "#s": "status" },
      ExpressionAttributeValues: {
        ":pk": { S: "u1" },
        ":st": { S: "OPEN" },
      },
    }),
  );
  assert.equal(out.Items?.length, 1);
  assert.equal(out.Items[0].pk.S, "u1");
  assert.equal(out.Items[0].status.S, "OPEN");
});

test("DynamoDB PartiQL ExecuteStatement", async (t) => {
  if (!(await requireReady(t))) return;
  const client = newDDB();
  const table = `${uniquePrefix()}-partiql`;

  await client.send(
    new CreateTableCommand({
      TableName: table,
      AttributeDefinitions: [{ AttributeName: "pk", AttributeType: "S" }],
      KeySchema: [{ AttributeName: "pk", KeyType: "HASH" }],
      BillingMode: "PAY_PER_REQUEST",
    }),
  );
  t.after(async () => {
    try {
      await client.send(new DeleteTableCommand({ TableName: table }));
    } catch {
      /* ignore */
    }
  });

  await client.send(
    new ExecuteStatementCommand({
      Statement: `INSERT INTO "${table}" VALUE {'pk':?,'data':?}`,
      Parameters: [{ S: "1" }, { S: "partiql" }],
    }),
  );
  const sel = await client.send(
    new ExecuteStatementCommand({
      Statement: `SELECT * FROM "${table}" WHERE pk=?`,
      Parameters: [{ S: "1" }],
    }),
  );
  assert.equal(sel.Items?.length, 1);
  assert.equal(sel.Items[0].data.S, "partiql");
});
